### PR TITLE
support Netty io_uring transport

### PR DESCRIPTION
**Pull request by sullis**
- https://en.wikipedia.org/wiki/Io_uring
- https://github.com/netty/netty-incubator-transport-io_uring

### CODE REVIEW

**Comment by sullis** [comment]
fyi @normanmaurer

**Comment by normanmaurer** [comment]
nice...!

**Comment by sullis** [comment]
Work in progress. cc: @agoniberg

**Comment by sullis** [comment]
Work in progress 
cc: @artgon

**Comment by sullis** [comment]
@argha-c  This is a draft PR of io_uring support. WDYT?

File: zuul-core/src/main/java/com/netflix/zuul/netty/server/Server.java [hunk 1]
```diff
@@ -378,6 +382,15 @@ private void initializeTransport()
                         workerExecutor,
                         chooserFactory,
                         DefaultSelectStrategyFactory.INSTANCE);
+            } else if (!useNio && ioUringIsAvailable()) {
```
[the remark this answers is not in the document]
Comment (argha-c) [reply zuul-core/src/main/java/com/netflix/zuul/netty/server/Server.java:385]: Also, I'd consider adding a toggle if the user wants to explicitly enable this.

File: zuul-io-uring-test/src/test/java/com/netflix/zuul/netty/server/IoUringTest.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+            assertNotEquals(inetAddress.getPort(), 0);
+            openHttpConnection(inetAddress.getPort());
+        });
+
+        s.stop();
+
+        assertEquals(2, ioUringChannelCount.get());
+    }
+
+    private static void openHttpConnection(final int port) {
```
Comment (argha-c) [inline zuul-io-uring-test/src/test/java/com/netflix/zuul/netty/server/IoUringTest.java:126]: This doesn't actually validate a proper request-response?

File: zuul-core/src/main/java/com/netflix/zuul/netty/server/Server.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (sullis) [reply zuul-core/src/main/java/com/netflix/zuul/netty/server/Server.java]: I just added FORCE_IO_URING to Server.java. Let me know what you think.

File: zuul-core/src/main/java/com/netflix/zuul/netty/server/Server.java [hunk 3]
```diff
@@ -354,8 +365,19 @@ private void initializeTransport()
             Executor workerExecutor = new ThreadPerTaskExecutor(workerThreadFactory);
 
             Map<ChannelOption<?>, Object> extraOptions = new HashMap<>();
-            boolean useNio = FORCE_NIO.get();
-            if (!useNio && epollIsAvailable()) {
+            final boolean useNio = FORCE_NIO.get();
+            final boolean useIoUring = FORCE_IO_URING.get();
+            if (useIoUring) {
```
Comment (artgon) [inline zuul-core/src/main/java/com/netflix/zuul/netty/server/Server.java:370]: Shouldn't this also check if it's available?

File: zuul-io-uring-test/src/test/java/com/netflix/zuul/netty/server/IoUringTest.java [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@RunWith(JUnit4.class)
+public class IoUringTest {
+    private static final boolean IS_OS_LINUX = "linux".equals(PlatformDependent.normalizedOs());
+    private static final Logger LOGGER = LoggerFactory.getLogger(IoUringTest.class);
+
+    @Test
+    public void testIoUringServer() throws Exception {
```
Comment (artgon) [inline zuul-io-uring-test/src/test/java/com/netflix/zuul/netty/server/IoUringTest.java:61]: What's the goal of this test? Just verifying that the server starts up?

File: zuul-io-uring-test/src/test/java/com/netflix/zuul/netty/server/IoUringTest.java (hunk 2 above)
Comment (sullis) [reply zuul-io-uring-test/src/test/java/com/netflix/zuul/netty/server/IoUringTest.java]: I replaced `openHttpConnection` with `checkConnection`

File: zuul-core/src/main/java/com/netflix/zuul/netty/server/Server.java (hunk 3 above)
Comment (sullis) [reply zuul-core/src/main/java/com/netflix/zuul/netty/server/Server.java]: Fixed.

File: zuul-io-uring-test/src/test/java/com/netflix/zuul/netty/server/IoUringTest.java (hunk 4 above)
Comment (sullis) [reply zuul-io-uring-test/src/test/java/com/netflix/zuul/netty/server/IoUringTest.java:72]: I just added a JavaDoc comment that explains the goal of the test.

**Comment by sullis** [comment]
I rebased this PR with the master branch.

Ready for review @argha-c

**Comment by artgon** [comment]
I think this is really cool, but seems like it's "incubator" mode so we're unlikely to run it in production (yet). I'm open to merging it to allow people to play around with it but it won't have the kind of verification that the rest of the code has.

**Comment by sullis** [comment]
> I think this is really cool, but seems like it's "incubator" mode so we're unlikely to run it in production (yet). I'm open to merging it to allow people to play around with it but it won't have the kind of verification that the rest of the code has.

Agreed.  It is an "incubator" feature.  I think it would be good to merge it so that folks can play around with it.

**Comment by sullis** [comment]
I resolved all merge conflicts.

**Comment by argha-c** [comment]
@sullis : Thanks for addressing the feedback so far. We're definitely open to merging it, and appreciate your contributions.

If you could consolidate and exercise the tests in our regular module (now that we have the FP), I think it'd be more valuable than introducing a separate test module altogether.

**Comment by sullis** [comment]
> @sullis : Thanks for addressing the feedback so far. We're definitely open to merging it, and appreciate your contributions.
> 
> If you could consolidate and exercise the tests in our regular module (now that we have the FP), I think it'd be more valuable than introducing a separate test module altogether.

I moved IoUringTest.java to zuul-core

Is this ok or should I move the test case into ServerTest.java ?

**Review by argha-c — APPROVED** [review]

**Comment by argha-c** [comment]
LGTM. Please rebase and resolve conflicts.

**Comment by sullis** [comment]
> LGTM. Please rebase and resolve conflicts.

Done.

**Pull request merged**